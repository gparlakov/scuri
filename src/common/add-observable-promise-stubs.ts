import { EOL } from 'os';
import { ConstructorParam, DependencyMethodReturnTypes } from '../types';

export function addDefaultObservableAndPromiseToSpyJoined(p: ConstructorParam, deps?: DependencyMethodReturnTypes, options?: DefaultMethodReturnsOptions): string {
    if(!deps?.has(p.type)) {
        return ''
    }
    const joiner = typeof options?.joiner === 'string' ? options?.joiner : EOL;
    
    return `${typeof joiner === 'string'? joiner : ''}${addDefaultObservableAndPromiseToSpy(p, deps, options).join(joiner)}`;
}

export function addDefaultObservableAndPromiseToSpy(p: ConstructorParam, deps?: DependencyMethodReturnTypes, options?: DefaultMethodReturnsOptions): string[] {
    if(!deps?.has(p.type)) {
        return []
    }
    const spyReturn = (v: string) => options?.spyReturnType === 'jest' ? `.mockReturnValue(${v})` : `.and.returnValue(${v})`;
    const dep = deps.get(p.type);
    const observables = Array.from(dep!.entries())
        .filter(([_, value]) => value.match(/Observable<|Subject</))
        .map(([key,]) => `${p.name}.${key}${spyReturn('EMPTY')}`);

    const promises = Array.from(dep!.entries())
        .map(([key, value]) => {
            return [key, value];
        })
        .filter(([_, value]) => value.match(/Promise</))
        .map(([key]) => `${p.name}.${key}${spyReturn('new Promise(res => {})')}`);
    return [...observables, ...promises];
}

interface DefaultMethodReturnsOptions {
    joiner?: string;
    spyReturnType?: 'jasmine' | 'jest';
}

export function listAllDefaultReturns(params: ConstructorParam[], deps?: DependencyMethodReturnTypes, options?: DefaultMethodReturnsOptions) {
    return params.flatMap(p => addDefaultObservableAndPromiseToSpy(p, deps, options))
}