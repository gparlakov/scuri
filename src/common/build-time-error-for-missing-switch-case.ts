export function buildErrorForMissingSwitchCase(_: never, description: string): never {
    throw new Error(description);
}
