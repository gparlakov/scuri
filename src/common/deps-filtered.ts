import {
    DependencyCallDescription,
    DependencyMethodReturnAndPropertyTypes,
} from '../types';

export function depsFiltered(
    deps: DependencyMethodReturnAndPropertyTypes | undefined,
    skip: (d: DependencyCallDescription) => boolean
): DependencyMethodReturnAndPropertyTypes | undefined {
    if (deps == null || !(deps instanceof Map)) {
        return deps;
    }
    // prettier-ignore
    return new Map(
        [...deps.entries()]
            .map(([key, value]) => ([
                key, new Map(
                    [...value.entries()].filter(([, dep]) => !skip(dep))
                )
            ])
        )
    );
}


export function dependenciesWrap(ds: DependencyMethodReturnAndPropertyTypes | undefined) {

    return {
        skip(skipWhere: (d: DependencyCallDescription) => boolean) {
            return depsFiltered(ds, skipWhere)
        }
    }
}
