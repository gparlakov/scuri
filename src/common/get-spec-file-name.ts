import { basename, join, normalize } from '@angular-devkit/core';
import { applyPathTemplate, FileEntry, TEMPLATE_FILENAME_RE } from '@angular-devkit/schematics';
import { ClassTemplateData } from '../types';

export function getSpecFilePathName(name: string) {
    return name.replace('.ts', '.spec.ts');
    // const normalizedName = normalize(name);
    // const ext = extname(basename(normalizedName));

    // return name.split(ext)[0] + '.spec' + ext;
}

export function getSpecFileCustomName(
    classData: Omit<ClassTemplateData, 'levenshtein' | 'params' | 'publicMethods' | 'setupMethods'>,
    templateName: string
): string | undefined {
    const templateFile: FileEntry = {
        path: basename(normalize(templateName)),
        content: Buffer.from(''),
    };

    const fileResult = applyPathTemplate(classData)(templateFile);

    const folderfyFn = folderfy(classData.folder);

    return typeof fileResult?.path === 'string'
        ? folderfyFn(
              fileResult.path
                  // remove template if any
                  .replace(TEMPLATE_FILENAME_RE, '')
          )
        : undefined;
}

function folderfy(folder: string) {
    return (fileName: string) => {
        return join(
            normalize(folder),
            fileName
                // remove the path as a possible duplicate
                .replace(folder, '')
        );
    };
}
