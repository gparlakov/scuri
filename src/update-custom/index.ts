import { normalize, strings } from '@angular-devkit/core';
import {
    applyContentTemplate,
    FileEntry,
    Rule,
    SchematicContext,
    Tree
} from '@angular-devkit/schematics';
import { EOL } from 'os';
import { getSpecFileCustomName } from '../common/get-spec-file-name';
import { paths } from '../common/paths';
import { describeSource } from '../common/read/read';
import { scuriTemplateMark, updateCustomTemplateCut } from '../common/scuri-custom-update-template';
import { ClassTemplateData, Description, isClassDescription } from '../types';

export type Options = {
    name: string;
    classTemplate: string;
};

export function updateCustom(o: Options): Rule {
    return (tree: Tree, context: SchematicContext) => {
        const template = tree.read(o.classTemplate)?.toString('utf8');
        const fileUnderTestContent = tree.read(o.name)?.toString('utf8');

        if (typeof fileUnderTestContent === 'string' && typeof template === 'string') {
            const r = describeSource(o.name, fileUnderTestContent);
            const topClass = Array.isArray(r) ? r[0] : <Description>{};
            if (isClassDescription(topClass)) {
                const { constructorParams, publicMethods } = topClass;
                const {
                    specFileName,
                    fileName: normalizedName,
                    folderPathNormal: folder
                } = paths(o.name);

                const classData: ClassTemplateData = {
                    ...strings,
                    className: topClass.name,
                    name: topClass.name,
                    specFileName,
                    normalizedName: normalizedName,
                    folder,
                    constructorParams: 'not available for update - please use `params`',
                    declaration:
                        'not available for update - specific for scuri original style testing',
                    builderExports:
                        'not available for update - builder is specific for scuri original style testing - the setup function with a builder',
                    params: constructorParams,
                    publicMethods: publicMethods,
                    shorthand: (normalizedName ?? '')[0].toLowerCase(),
                }

                const customSpecFileName = getSpecFileCustomName(classData, o.classTemplate);
                const customSpecFileContents = customSpecFileName != null ? tree.read(customSpecFileName) : null;
                if (customSpecFileContents != null) {
                    const inserts = updateWithCustomTemplate(
                        {
                            ...classData,
                            specFileContents: customSpecFileContents.toString(
                                'utf8'
                            ),
                            templateContents: template,
                        },
                        context
                    );

                    const recorder = tree.beginUpdate(customSpecFileName!);
                    inserts.forEach((a) => {
                        recorder.insertLeft(a.position, a.content);
                    });
                    tree.commitUpdate(recorder);

                } else {
                    context.logger.error(`Looks like the custom spec file ${customSpecFileName} does not exist. Try creating the file before updating it.`)
                }
            } else {
                context.logger.error(`Looks like there was no class in ${o.name}`);
            }
        } else {
            if (typeof fileUnderTestContent !== 'string') {
                context.logger.error(`Looks like the file ${o?.name} is missing or invalid.`);
            }
            if (typeof template !== 'string') {
                context.logger.error(`Looks like the file ${o?.classTemplate} is missing or invalid.`);
            }
        }
        return tree;
    };
}

const skip = '-skipDeDupe';
// todo - check for each error condition and report
export function updateWithCustomTemplate(
    templateData: ClassTemplateData & { templateContents: string; specFileContents: string },
    context: SchematicContext
): { position: number; content: string }[] {
    const [_, parts] = updateCustomTemplateCut(templateData.templateContents);
    context.logger.debug(`Cut the template to parts ${JSON.stringify(parts)}`);
    if(parts == null || parts.length === 0) {
        throw new Error(`The custom template seems to be missing the ${scuriTemplateMark} mark. Perhaps you need the standard update?`)
    }
    
    const originalMethods = templateData.publicMethods;
    // skip methods that already have tests
    templateData.publicMethods = originalMethods.filter(
        (pm) => !templateData.specFileContents.includes(pm)
    );

    if (templateData.publicMethods.length !== originalMethods.length) {
        context.logger.debug(
            `Skipping methods: [${originalMethods
                .filter((m) => !templateData.publicMethods.includes(m))
                .join(',')}] as they seem to be already in the spec.`
        );
    }

    // curry (e.g. prepare) the template fn with the context (templateData)
    const templateFn = applyContentTemplate(templateData);

    return parts
        .map((p) => {
            const mark = `// scuri:${p.mark.replace(new RegExp(skip, 'i'), '')}`.toLowerCase();
            const template: FileEntry = {
                path: normalize('.'),
                content: Buffer.from(p.template),
            };
            const templateResult = templateFn(template)?.content.toString('utf8') ?? '';

            if (templateResult === '') {
                context.logger.debug(`No result from applying template for ${p.mark}.`);
            } else {
                context.logger.debug(`Template result before de-duplication: [${templateResult}]`);
            }
            
            const spaces = getWhitespaceBefore(templateData.specFileContents, mark);
            // skip de-duplication for the whole section (mark)
            const skipDeDupe = p.mark.toLowerCase().includes(skip.toLowerCase());
            const deDupedContent = templateResult
                .split(/\r\n|\r|\n/g)
                // remove lines of code that are already in the spec
                .filter((c) => skipDeDupe || !templateData.specFileContents.includes(c))
                .join(`${EOL}${spaces}`);

            const position = templateData.specFileContents.toLowerCase().indexOf(mark);
            context.logger.debug(`Mark ${mark} (original ${p.mark}) found at position(${position})`)
            // add the space at the end
            return { position, content: `${deDupedContent}${EOL}${spaces}` };
        })
        .filter((a) => !templateData.specFileContents.includes(a.content));
}
function getWhitespaceBefore(content: string, mark: string) {
    const spacesOrEmpty = content.match(new RegExp(`^.*${mark}`, 'gm')) ?? [''];
    return spacesOrEmpty[0].replace(mark, '');
}
