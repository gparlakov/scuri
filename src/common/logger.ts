import { Logger } from '@angular-devkit/core/src/logger';

let mainLogger: Logger
export function setLogger(l: Logger) {
    if(mainLogger != null) {
        // logger already set - don't overwrite
        return;
    }
    mainLogger = l;
}

export function getLogger(name: string): Logger {
    if(mainLogger == null) {
        setLogger(new Logger('just a logger'));
    }
    return mainLogger.createChild(name);
}
