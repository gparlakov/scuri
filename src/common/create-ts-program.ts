import { Tree } from '@angular-devkit/schematics';
import * as ts from 'typescript';
import { getLogger } from './logger';

let programCache = new Map<string, ts.Program>();
export function createTsProgram(fileName: string, tree: Tree): ts.Program {
    const logger = getLogger(createTsProgram.name);
    logger.debug(`entering for ${fileName}`)

    if(programCache.has(fileName)) {
        logger.debug(`Reusing program for ${fileName}`)
        return programCache.get(fileName)!;
    }

    let config: ts.ParsedCommandLine = {
        options: {
            target: ts.ScriptTarget.ES2015,
            module: ts.ModuleKind.CommonJS
        },
        fileNames: [],
        errors: [],
    };

    const closestTsConfig = ts.findConfigFile(fileName, (f) => tree.exists(f));
    if (typeof closestTsConfig === 'string') {
        logger.debug(`Found ${closestTsConfig}`);
        const src = ts.readJsonConfigFile(closestTsConfig, (f) => {
            if (tree.exists(f)) {
                const raw = tree.read(f);
                if (raw != null) {
                    return raw.toString();
                }
            }
        });
        logger.debug(`Parsed json ${src?.getText() ?? 'TsConfigSourceFile is nullish'}`);
        config = ts.parseJsonSourceFileConfigFileContent(
            src,
            {
                fileExists: (f) => {
                    logger.debug(`parseConfig: fileExists ${f}?${tree.exists(f)}`);
                    return tree.exists(f);
                },
                useCaseSensitiveFileNames: true,
                readFile: (f) => {
                    logger.debug(`parseConfig: readFile ${f} : ${tree.read(f)?.toString()}`);
                    return tree.read(f)?.toString();
                },
                readDirectory: (r, _exs, _excludes, _includes, _depth) => {
                    logger.debug(`parseConfig:readDirectory ${r}`);
                    return tree.getDir(r).subfiles.map((s) => s);
                },
            },
            tree.root.path,
            undefined,
            closestTsConfig
        );

        logger.debug(`parsed config: ${JSON.stringify(config)}`);
    } else {
        let checkedFiles: string[] = [];
        ts.findConfigFile(fileName, (f) => (checkedFiles.push(f), tree.exists(f)));
        logger.warn(
            `Did not find a tsconfig file close to ${fileName}. Some types might be missing. Searched in ${checkedFiles.join(
                '||'
            )}`
        );
    }

    const defaultHost = ts.createCompilerHost({}, true);
    logger.debug(
        `Creating (and caching) program for ${fileName}, with options\n${JSON.stringify(config.options)}`
    );
    const program = ts.createProgram([fileName], config.options, defaultHost);
    logger.debug(`FileContents: ${program.getSourceFile(fileName)?.getFullText()}`)
    programCache.set(fileName, program)
    return program;
}

export function getSourceFile(fullFileName: string, tree: Tree): ts.SourceFile | undefined {
    const logger = getLogger(getSourceFile.name);
    logger.debug(`entering for ${fullFileName}`)
    // const program = createTsProgram(fullFileName, tree);

    // if(program == null) {
    //     logger.debug('Program/source file is still null, probably some error in log above. Creating a new source file from scratch.')
    //     const contents = tree.read(fullFileName)?.toString();
    //     if(contents == null) {
    //         logger.error(`The contents of the ${fullFileName} are empty - the file is probably missing. Can't create a source file`)
    //         return undefined;
    //     }
    //     logger.debug(`Creating just a source file for ${fullFileName}`)
    return ts.createSourceFile(fullFileName, tree.read(fullFileName)?.toString() ?? '', ts.ScriptTarget.Latest, true);
    // }

    // logger.debug('Found program. Returning the source file from it')
    // return program.getSourceFile(fullFileName);
}
// failed experiment - faking the ts host by reading/writing from the tree host
// failed b/c tree does not get node_modules and other  necessary .d.ts files that comprise a big chunk of the base types like string, Number, HTMLElement, etc.
// keeping it around as it took way too much effort to just delete... :(

// export function createTsProgram_1(fileName: string): ts.Program {
//     const logger = getLogger(createTsProgram.name);

//     let config: ts.ParsedCommandLine = {
//         options: {
//             target: ts.ScriptTarget.ES2015,
//             module: ts.ModuleKind.CommonJS,
//         },
//         fileNames: [],
//         errors: [],
//     };

//     const closestTsConfig = ts.findConfigFile(fileName, (f) => tree.exists(f));
//     if (typeof closestTsConfig === 'string') {
//         logger.debug(`Found ${closestTsConfig}`);
//         const src = ts.readJsonConfigFile(closestTsConfig, (f) => {
//             if (tree.exists(f)) {
//                 const raw = tree.read(f);
//                 if (raw != null) {
//                     return raw.toString();
//                 }
//             }
//         });
//         logger.debug(`Parsed json ${src?.getText() ?? 'TsConfigSourceFile is nullish'}`);
//         config = ts.parseJsonSourceFileConfigFileContent(
//             src,
//             {
//                 fileExists: (f) => {
//                     logger.debug(`parseConfig: fileExists ${f}?${tree.exists(f)}`);
//                     return tree.exists(f);
//                 },
//                 useCaseSensitiveFileNames: true,
//                 readFile: (f) => {
//                     logger.debug(`parseConfig: readFile ${f} : ${tree.read(f)?.toString()}`);
//                     return tree.read(f)?.toString();
//                 },
//                 readDirectory: (r, _exs, _excludes, _includes, _depth) => {
//                     logger.debug(`parseConfig:readDirectory ${r}`);
//                     return tree.getDir(r).subfiles.map((s) => s);
//                 },
//             },
//             tree.root.path,
//             undefined,
//             closestTsConfig
//         );

//         logger.debug(`parsed config: ${JSON.stringify(config)}`);
//     } else {
//         let checkedFiles: string[] = [];
//         ts.findConfigFile(fileName, (f) => (checkedFiles.push(f), tree.exists(f)));
//         logger.warn(
//             `Did not find a tsconfig file close to ${fileName}. Some types might be missing. Searched in ${checkedFiles.join(
//                 '\n'
//             )}`
//         );
//     }

//     const defaultHost = ts.createCompilerHost({}, true);
//     const treeHost = <ts.CompilerHost>{
//         ...defaultHost,
//         readFile: (s: string) => {
//             const res = tree.read(s)?.toString();
//             logger.debug(`treeHost: readFile ${s}: ${res}`);
//             return res;
//         },
//         writeFile: (name: string, text: string) => {
//             logger.debug(`treeHost: writing ${name}:${text}`);
//             tree.overwrite(name, text);
//         },
//         fileExists: (name: string) => {
//             logger.debug(`treeHost: exists ${name}:${tree.exists(name)}`);
//             return tree.exists(name);
//         },
//         getNewLine: () => EOL,
//         getCurrentDirectory: () => {
//             logger.debug(`treeHost: getCurrentDirectory ${tree.root.path}}`);
//             return tree.root.path
//         },
//         getDirectories: (path: string) => {
//             logger.debug(`treeHost: getDirectories for ${path}: ${tree.getDir(path)?.subdirs}`);
//             return tree.getDir(path)?.subdirs
//         },
//         directoryExists: (directoryName: string) => {
//             logger.debug(`treeHost: directoryExists for ${directoryName}: ${tree.getDir(directoryName) != null}`);
//             return tree.getDir(directoryName) != null
//         },
//         readDirectory(rootDir, extensions, excludes, includes, depth?) {
//             logger.debug(`treeHost: readDirectory for, ${rootDir}, ${extensions}, ${excludes}, ${includes}, ${depth}`);
//             return tree.getDir(rootDir).subfiles
//         },
//     };
//     logger.debug(
//         `Creating program for ${fileName}, with options\n${JSON.stringify(config.options)}`
//     );
//     return ts.createProgram([fileName], config.options, treeHost);
// }


