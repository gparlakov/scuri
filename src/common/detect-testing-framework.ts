import { Tree } from '@angular-devkit/schematics';
import { getLogger } from './logger';

export type Supported = 'jest' | 'jasmine';
const supported: Supported[] = ['jest', 'jasmine'];

/**
 * Take the framework from the 1. cli params or 2. config or 3. best effort autodetect
 * @param cliParam command line interface param if any
 * @param configParam configured param if any
 * @param t Tree
 * @param fallback - if all else fails use this
 * @returns 'jest' | 'jasmine'
 */
export function detectTestingFramework(
    cliParam: string | undefined,
    configParam: string | undefined,
    t: Tree,
    fallback: Supported
): 'jest' | 'jasmine' {
    const logger = getLogger(detectTestingFramework.name);

    logger.log('debug', 'entering');
    if (typeof cliParam === 'string' && supported.includes(cliParam as Supported)) {
        logger.debug(`Detected supported cli param ${cliParam}`);
        return cliParam as Supported;
    }
    if (typeof cliParam === 'string') {
        logger.warn(
            `Detected un-supported cli param ${cliParam}. please use ${supported.join(' or ')}`
        );
    }
    if (typeof configParam === 'string' && supported.includes(configParam as Supported)) {
        logger.debug(`Detected supported config param ${cliParam}`);
        return configParam as Supported;
    }
    if (typeof configParam === 'string') {
        logger.warn(
            `Detected un-supported config param ${configParam}. please use ${supported.join(
                ' or '
            )}`
        );
    }

    const found: {
        jest: string[];
        jasmine: string[];
    } = { jest: [], jasmine: [] };

    const root = t.getDir('.');

    root.subfiles.forEach(f => {
        if(f.includes('node_modules')) {
            // skip node modules
            return;
        }

        if (f.includes('jest')) {
            found.jest.push(f);
            logger.debug(`Detected jest by way of ${f}`);
        } else if (f.includes('karma')) {
            found.jasmine.push(f);
            logger.debug(`Detected jasmine by way of: ${f}`);
        } else if (f.includes('package.json')) {
            const content = t.read(`./${f}`)?.toString();
            if (typeof content === 'string') {
                if (content.includes('"jasmine-core"')) {
                    found.jasmine.push(f);
                    logger.debug(`Detected jasmine by way of ${f}`);
                }
                if (content.includes('"jest"')) {
                    found.jest.push(f);
                    logger.debug(`Detected jest by way of ${f}`);
                }
            }
        }
    });

    if (found.jasmine.length > found.jest.length) {
        logger.debug(`Detected jasmine by way of ${found.jasmine.join()}`);
        return 'jasmine';
    } else if (found.jest.length > found.jasmine.length) {
        logger.debug(`Detected jest by way of ${found.jest.join()}`);
        return 'jest';
    } else {
        logger.debug(`Fallback to ${fallback}`);
        // jest == jasmine
        return fallback;
        // todo - read through the files again and get the .mockReturnValue  and '.and.' numbers and .returnValue counts to make sure which type of runner is being used
    }
}
