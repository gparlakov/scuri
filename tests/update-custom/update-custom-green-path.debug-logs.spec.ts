import { Tree } from '@angular-devkit/schematics';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { Options } from '../../src/update-custom/index';
import { listenLogger, subscribe } from '../get-logger-errors';
import { collectionPath } from '../spec/common';
describe('update-custom', () => {
    let tree: Tree;
    const name = 'example.ts';
    const specFileName = 'example.spec.ts';
    const classTemplate = '__specFileName__.template';

    beforeEach(() => {
        tree = Tree.empty();

        tree.create(name, fileContent());
        tree.create(specFileName, specFileContent());
        tree.create(classTemplate, template());
    });

    it('should debug-output the skipped methods and the template results prior to dedupe', async () => {
        const runner = new SchematicTestRunner('schematics', collectionPath);
        const logs = subscribe(listenLogger(runner.logger, { level: 'debug' }), 5); // expecting 5 entries

        await runner
            .runSchematicAsync('update-custom', <Options>{ name, classTemplate, specFileContents: specFileContent() }, tree)
            .toPromise();

        // skipping methods
        expect(logs[0]).toEqual('Skipping methods: [myMethod] as they seem to be already in the spec.');

        // dedupe lets
        expect(logs[1]).toEqual(`Template result before de-duplication: [let serviceSpy: Service;
let routerSpy: Router;
let justSpy: Just;
]`);

        // dedupe providers
        expect(logs[2]).toEqual(`Template result before de-duplication: [{ provide: Service, useClass: autoSpy(Service, 'Service') },
{ provide: Router, useClass: autoSpy(Router, 'Router') },
{ provide: Just, useClass: autoSpy(Just, 'Just') },
]`);

        // dedupe instances
        expect(logs[3]).toEqual(`Template result before de-duplication: [serviceSpy = spyInject<Service>(TestBed.inject(Service));
routerSpy = spyInject<Router>(TestBed.inject(Router));
justSpy = spyInject<Just>(TestBed.inject(Just));
]`);

        // dedupe methods
        expect(logs[4]).toEqual(`Template result before de-duplication: [it('when yourMethod is called it should', () => {
    // arrange
    // act
    e.yourMethod();
    // assert
    // expect(e).toEqual
});
it('when theirMethod is called it should', () => {
    // arrange
    // act
    e.theirMethod();
    // assert
    // expect(e).toEqual
});
]`);
    });

    it('should debug-output the empty template results', async () => {
        const r = tree.beginUpdate(classTemplate)
        r.insertRight(1813, '/**scuri:template:empty-template:*/');
        tree.commitUpdate(r);
        const runner = new SchematicTestRunner('schematics', collectionPath);
        const logs = subscribe(listenLogger(runner.logger, { level: 'debug' }), 6); // expecting 5 entries

        await runner
            .runSchematicAsync('update-custom', <Options>{ name, classTemplate, specFileContents: specFileContent() }, tree)
            .toPromise();

        expect(logs[5]).toEqual('No result from applying template for empty-template.');
    });

    it('should error on failed template', async () => {
        const r = tree.beginUpdate(classTemplate)
        r.insertRight(1813, '/**scuri:template:empty-template:<%= test.forEach %> <% sharans.forEach(s => {%> tt<%=s <% }%>*/');
        tree.commitUpdate(r);
        const runner = new SchematicTestRunner('schematics', collectionPath);
        const logs = subscribe(listenLogger(runner.logger, { level: 'error' }), 5); // expecting 5 entries

        await runner
            .runSchematicAsync('update-custom', <Options>{ name, classTemplate, specFileContents: specFileContent() }, tree)
            .toPromise()
            .catch(e => {
                expect(logs).toEqual([]);
                expect(e).toEqual(new SyntaxError('Unexpected token \'%\''));
            })
    });
});

function fileContent(): string {
    return `import { Router } from "@the/router";
import { Just } from "maybe";
import { Service } from "./service";
export class ExampleComponent {
    constructor(service: Service, router: Router, just: Just) {}

    myMethod() {}

    yourMethod() {}

    theirMethod() {}
}
`;
}

function specFileContent(): string {
    return `import { MyDirective } from './directive';
import { autoSpy, spyInject } from 'jasmine-auto-spies';

describe('ExampleComponent', () => {
    // scuri:lets

    beforeEach(
        waitForAsync(() => {
            TestBed.configureTestingModule({
                providers: [
                    MyDirective,
                    { provide: Service, useClass: autoSpy(Service, 'Service') },
                    // scuri:injectables
                ]
            });

            directive = TestBed.inject(MyDirective);
            // scuri:get-instances
        })
    );

    it('when myMethod is called it should', () => {
        // arrange
        // act
        e.myMethod();
        // assert
        // expect(e).toEqual
    });

    // scuri:methods
});
`;
}

function template() {
    return `<% params.forEach(p => { if(p.importPath) {%>import { <%= p.type %> } from '<%= p.importPath %>';
<% }}) %>import { <%= className %> } from './<%= normalizedName %>';
import { autoSpy, spyInject } from 'jasmine-auto-spies';

describe('<%= className %>', () => {
  <%params.forEach(p => {%> let <%= p.type %>Spy: <%= p.type %>;
      <% }) %>
    // scuri:lets

  beforeEach(
  waitForAsync(() => {
      TestBed.configureTestingModule({
      providers: [
              MyDirective,
          <%params.forEach(p => {%> { provide: <%= p.type %>, useClass: autoSpy(<%= p.type %>, '<%= p.type %>') },
          <% }) %>
          // scuri:injectables
      ]
      });

      directive = TestBed.inject(MyDirective);
      <%params.forEach(p => {%> <%= p.type %>Spy = spyInject<<%= p.type %>>(TestBed.inject(<%= p.type %>));
      <% }) %>
      // scuri:get-instances

  })
  );

  <% publicMethods.forEach(meth=> {if(meth != '') { %>it('when <%= meth %> is called it should', () => {
  // arrange
  // act
  <%= shorthand %>.<%= meth %>();
  // assert
  // expect(<%= shorthand %>).toEqual
  });
  <% }}) %>

  // scuri:methods
});

/** scuri:template:lets:<%params.forEach(p => {%>let <%= camelize(p.type) %>Spy: <%= p.type %>;
<% }) %>*/
/** scuri:template:injectables:<%params.forEach(p => {%>{ provide: <%= p.type %>, useClass: autoSpy(<%= p.type %>, '<%= p.type %>') },
<% }) %>*/
/** scuri:template:get-instances:<%params.forEach(p => {%><%= camelize(p.type) %>Spy = spyInject<<%= p.type %>>(TestBed.inject(<%= p.type %>));
<% }) %>*/
/** scuri:template:methods-skipDeDupe:<% publicMethods.forEach(meth=> {if(meth != '') { %>it('when <%= meth %> is called it should', () => {
    // arrange
    // act
    <%= shorthand %>.<%= meth %>();
    // assert
    // expect(<%= shorthand %>).toEqual
});
<% }}) %>*/
`;
}
