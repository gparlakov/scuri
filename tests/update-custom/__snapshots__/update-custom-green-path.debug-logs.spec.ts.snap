// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update-custom should debug-output the skipped methods and the template results prior to dedupe 1`] = `
Array [
  "Entering for opts:{\\"name\\":\\"C:\\\\\\\\Users\\\\\\\\gparl\\\\\\\\projects\\\\\\\\scuri\\\\\\\\tests\\\\\\\\spec\\\\\\\\test-data\\\\\\\\debug-logs\\\\\\\\for-debug-logs.ts\\",\\"update\\":true,\\"classTemplate\\":\\"__specFileName__.template\\"}",
  "entering",
  "entering for C:\\\\Users\\\\gparl\\\\projects\\\\scuri\\\\tests\\\\spec\\\\test-data\\\\debug-logs\\\\for-debug-logs.ts",
  "Creating (and caching) program for C:\\\\Users\\\\gparl\\\\projects\\\\scuri\\\\tests\\\\spec\\\\test-data\\\\debug-logs\\\\for-debug-logs.ts, with options
{\\"target\\":2,\\"module\\":1}",
  "Found src file: (top 50) import { Router } from \\"@the/router\\";
import { Ju...",
  "entering for C:\\\\Users\\\\gparl\\\\projects\\\\scuri\\\\tests\\\\spec\\\\test-data\\\\debug-logs\\\\for-debug-logs.ts",
  "returning {\\"specFileName\\":\\"for-debug-logs.spec.ts\\",\\"fileName\\":\\"for-debug-logs\\",\\"folderPathRaw\\":\\"C:\\\\\\\\Users\\\\\\\\gparl\\\\\\\\projects\\\\\\\\scuri\\\\\\\\tests\\\\\\\\spec\\\\\\\\test-data\\\\\\\\debug-logs\\\\\\\\\\",\\"folderPathNormal\\":\\"C:\\\\\\\\Users\\\\\\\\gparl\\\\\\\\projects\\\\\\\\scuri\\\\\\\\tests\\\\\\\\spec\\\\\\\\test-data\\\\\\\\debug-logs\\\\\\\\\\"}",
  "entering for classData {\\"className\\":\\"ExampleComponent\\",\\"name\\":\\"ExampleComponent\\",\\"specFileName\\":\\"for-debug-logs.spec.ts\\",\\"normalizedName\\":\\"for-debug-logs\\",\\"folder\\":\\"C:\\\\\\\\Users\\\\\\\\gparl\\\\\\\\projects\\\\\\\\scuri\\\\\\\\tests\\\\\\\\spec\\\\\\\\test-data\\\\\\\\debug-logs\\\\\\\\\\",\\"constructorParams\\":\\"not available for update - please use \`params\`\\",\\"declaration\\":\\"not available for update - specific for scuri original style testing\\",\\"builderExports\\":\\"not available for update - builder is specific for scuri original style testing - the setup function with a builder\\",\\"params\\":[{\\"name\\":\\"service\\",\\"type\\":\\"Service\\",\\"importPath\\":\\"./service\\"},{\\"name\\":\\"router\\",\\"type\\":\\"Router\\",\\"importPath\\":\\"@the/router\\"},{\\"name\\":\\"just\\",\\"type\\":\\"Just\\",\\"importPath\\":\\"maybe\\"}],\\"publicMethods\\":[\\"myMethod\\",\\"yourMethod\\",\\"theirMethod\\"],\\"shorthand\\":\\"f\\",\\"setupMethods\\":[]} templateName __specFileName__.template",
  "looking for templateFile __specFileName__.template result in [object Object]",
  "returning C:\\\\Users\\\\gparl\\\\projects\\\\scuri\\\\tests\\\\spec\\\\test-data\\\\debug-logs\\\\for-debug-logs.spec.ts",
  "/__specFileName__.template",
  "/C:/Users/gparl/projects/scuri/tests/spec/test-data/debug-logs/for-debug-logs.ts",
  "/C:/Users/gparl/projects/scuri/tests/spec/test-data/debug-logs/for-debug-logs.spec.ts",
  "Cut the template to parts [{\\"mark\\":\\"lets\\",\\"template\\":\\"<%params.forEach(p => {%>let <%= camelize(p.type) %>Spy: <%= p.type %>;\\\\n<% }) %>\\"},{\\"mark\\":\\"injectables\\",\\"template\\":\\"<%params.forEach(p => {%>{ provide: <%= p.type %>, useClass: autoSpy(<%= p.type %>, '<%= p.type %>') },\\\\n<% }) %>\\"},{\\"mark\\":\\"get-instances\\",\\"template\\":\\"<%params.forEach(p => {%><%= camelize(p.type) %>Spy = spyInject<<%= p.type %>>(TestBed.inject(<%= p.type %>));\\\\n<% }) %>\\"},{\\"mark\\":\\"methods-skipDeDupe\\",\\"template\\":\\"<% publicMethods.forEach(meth=> {if(meth != '') { %>it('when <%= meth %> is called it should', () => {\\\\n    // arrange\\\\n    // act\\\\n    <%= shorthand %>.<%= meth %>();\\\\n    // assert\\\\n    // expect(<%= shorthand %>).toEqual\\\\n});\\\\n<% }}) %>\\"}]",
  "Skipping methods: [myMethod] as they seem to be already in the spec.",
  "Template result before de-duplication: [let serviceSpy: Service;
let routerSpy: Router;
let justSpy: Just;
]",
  "Mark // scuri:lets (original lets) found at position(146)",
  "Template result before de-duplication: [{ provide: Service, useClass: autoSpy(Service, 'Service') },
{ provide: Router, useClass: autoSpy(Router, 'Router') },
{ provide: Just, useClass: autoSpy(Just, 'Just') },
]",
  "Mark // scuri:injectables (original injectables) found at position(422)",
  "Template result before de-duplication: [serviceSpy = spyInject<Service>(TestBed.inject(Service));
routerSpy = spyInject<Router>(TestBed.inject(Router));
justSpy = spyInject<Just>(TestBed.inject(Just));
]",
  "Mark // scuri:get-instances (original get-instances) found at position(548)",
  "Template result before de-duplication: [it('when yourMethod is called it should', () => {
    // arrange
    // act
    f.yourMethod();
    // assert
    // expect(f).toEqual
});
it('when theirMethod is called it should', () => {
    // arrange
    // act
    f.theirMethod();
    // assert
    // expect(f).toEqual
});
]",
  "Mark // scuri:methods (original methods-skipDeDupe) found at position(770)",
]
`;
